providers/gitlab.go

```diff
@@ -213,8 +256,37 @@ func (p *GitLabProvider) EnrichSessionState(ctx context.Context, s *sessions.Ses
 		return fmt.Errorf("group membership check failed: %v", err)
 	}
 
+	p.addProjectMembershipToState(ctx, s)
+
 	s.User = userInfo.Username
 	s.Email = userInfo.Email
 
 	return nil
+
+}
+
+// addProjectMembership adds projects into session.Groups
+func (p *GitLabProvider) addProjectMembershipToState(ctx context.Context, s *sessions.SessionState) {
+	if len(p.Projects) == 0 {
+		return
+	}
+
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, project := range p.Projects {
+		project, err := p.getProjectInfo(ctx, s, project)
+		// If no error, it's a member of the project
+		if err == nil && !project.Archived {
+			s.Groups = append(s.Groups, project.PathWithNamespace)
```

Comment:
Since this provider will potentially have multiple group-like things in the `session.Groups`, we need to scope them to avoid collisions.

@JoelSpeed and I have talked about this in the past and think this structure is best:

`project:<PROJECT>`

Also, why the Path and not the name? Will name alone have collision issues?

Reply:
If the path is always `group/project` structure, you might be fine just using the project name & relying on the use specifying `-gitlab-groups` for that authz restriction.

BUT - I think this opens up a scenario where we need to think if Gitlab should override the default `Authorize` implementation that just does set intersection logic:

If both `gitlab-group` and `gitlab-project` are set, do you want any being valid to suffice? Or do you want the user to be in at least one of the listed groups AND at least one of the listed projects.

If that's the case, we need to override `Authorize` here and check each of those buckets for membership distinctly.

Reply:
You could also make these exclusive options and only allow a configuration to set 1 or the other for authZ logic? I'm not a Gitlab user, so I don't know what makes sense for how users would configure it.

Reply:
Thanks for the clarification about the prefix, was unsure if it was planned/discussed or if I have to implement this right away.

> Also, why the Path and not the name? Will name alone have collision issues?

The full path is the only way to point to a unique repository name. Since Gitlab 9.0, the notion of "subgroup" is available so you can have something like `group/sub-group/project` and much more since it support up to 20 level of subgroups.

It will avoid potential issues or surprises for people using this feature.

> If both gitlab-group and gitlab-project are set, do you want any being valid to suffice? Or do you want the user to be in at least one of the listed groups AND at least one of the listed projects.

For me, groups and projects are different entities. Project scope is thinner than group scope, for example, you can have access to a project without being a group member of the owner group of the target project.

For now, groups works with an **OR** combination. If you list multiple group, you need to be a member of at least one of them. Following this, I think we should do the same with projects : if you have a combination of projects and groups, you need to be part of the group or the project. The simpler the better.

While writing this answer, I thought that adding a `gitlab-project-access-level` option will be better for fine tuning. In a project you can have multiple roles, as described in the [documentation](https://docs.gitlab.com/ee/api/members.html#valid-access-levels). I think by default we should set the `reporter` level, and use a string to specify custom access level. I do not know if this is implemented in other providers but I think it's a nice and clean feature. What do you think ?

Reply:
Might also be useful to potentially look at the Github provider as a reference between the 3 entities they do AuthZ with. It looks like they have interesting combination of logic between the 3 types of entities for AuthZ.

Seems like both pools continue to be OR logic within them. But we need to either make both being set not allowed (so a validation check) OR if both are set, require membership in both sets of entities.

Reply:
@JoelSpeed Just an FYI - this circles back to our discussion of if `Authorize` made sense to be a provider method at all that could be override-able. Looking at the logic in Github, it seems like some providers might have scenarios where they have complex logic between group types where a global OR logic wouldn't suffice.